### PR TITLE
Format code to almost comply with ruff / flake8

### PR DESCRIPTION
**Pull request by steinmig**
There are still some open issues that ruff and flake8 complain about, but these require more detailed attention in a separate PR.

As the code is not completely clean for Ruff / Flake8 / mypy yet, these are not added to the CI yet.

### CODE REVIEW

File: nff/utils/geom.py [hunk 1]
```diff
@@ -216,64 +214,49 @@ def compute_distances(dataset,
...
+            atom_nxyz = batch_1["nxyz"].reshape(num_mols_1, -1, 4).to(device)
 
-            out = compute_distance(targ_nxyz=targ_nxyz,
-                                   atom_nxyz=atom_nxyz,
-                                   store_grad=store_grad)
+            out = compute_distance(targ_nxyz=targ_nxyz, atom_nxyz=atom_nxyz, store_grad=store_grad)
             if store_grad:
                 distances, R, xyz_0 = out
-                num_atoms = batch_1["num_atoms"].tolist()
+                batch_1["num_atoms"].tolist()
```
Comment (ajhoffman1229) [inline nff/utils/geom.py:236]: probably delete unused assignment

File: nff/io/bias_calculators.py
Comment (ajhoffman1229) [inline nff/io/bias_calculators.py:1]: I have also made several changes to this in a separate branch, adding documentation and reformatting. I can review more closely when I merge my version.

File: nff/train/uncertainty.py
Comment (ajhoffman1229) [inline nff/train/uncertainty.py:1]: I believe this is the only other file where I have made changes in my branch. Again, I can review it more closely when I merge my version (but I suspect my changes will supersede any formatting changes made in a first-pass of ruff formatting).

**Review by ajhoffman1229 — CHANGES_REQUESTED** [review]
Thanks so much for doing this formatting pass, Miguel! I have a few very minor comments about some changes to imports that ruff made, questions about changes to the `pyproject.toml` file for the repo, and some small suggestions for deleting unneeded code that ruff identified. I think most of my comments do not require any edits to the code, but I would like to make sure that I understand all the changes (esp. to the TOML) before merging.

There was at least one file that had imports from `htvs` and that called `django.setup()`. I commented on that part, but I can't find it now. It might be a file that we need to update or remove entirely. In principle, NFF should be totally separate from `htvs` (especially as an external repo, we don't want to have any dependencies on our private, internal repo that risks breaking any functionality for users who aren't in our group).

Also, there are some files that I'm working on in a separate branch that I have already updated to abide by ruff's formatting rules and the linting rules in `pyproject.toml`, along with adding significant documentation. I've noted which files will need additional consideration when I merge my changes.

File: nff/md/ci/opt.py [hunk 2]
```diff
@@ -28,40 +15,30 @@
...
+
+
+os.environ["DJANGO_SETTINGS_MODULE"] = "djangochem.settings.orgel"
+django.setup()
+
+# Shell Plus Model Imports
+
+
+# from nff.nn.models import PostProcessModel
+
```
[the remark this answers is not in the document]
Comment (ajhoffman1229) [reply nff/md/ci/opt.py:41]: Actually, after looking at this more closely, I am inclined to say that it should be deleted entirely. I think this is a test of one of Simon's models, but it's not obvious from the code what it's doing (I presume he had some configuration interaction calculations that he used to train a NFF?), it refers to a repo on which NFF should not depend at all, and has lots of broken code (missing/commented out imports, references to Simon's specific fork of `htvs`, references to Simon's personal folders on a workstation, etc.). It hasn't been updated in nearly 4 years and, from what I can tell, @xiaochendu was in the process of [deprecating it](https://github.com/learningmatter-mit/NeuralForceField/commit/aa24cd513178f24391bc06a6a00642d864517a5a) anyway. If @xiaochendu knows more, it would be great if he could chime in here!

File: nff/io/cprop.py [hunk 3]
```diff
@@ -25,15 +23,11 @@ def get_cp_cmd(script,
       cmd (str): the chemprop command
     """
 
-    cmd = (f"python {script} --config_path {config_path} "
-           f" --data_path {data_path} "
-           f" --dataset_type {dataset_type}")
+    cmd = f"python {script} --config_path {config_path} " f" --data_path {data_path} " f" --dataset_type {dataset_type}"
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/io/cprop.py:26]: Well spotted, thanks!

```suggestion
    cmd = f"python {script} --config_path {config_path}  --data_path {data_path}  --dataset_type {dataset_type}"
```

File: nff/md/ci/opt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/md/ci/opt.py:41]: Yes, that file was on my list for the next PR for sure. Deleting it sounds good to me, apparently nobody besides Simon has ever imported it in the last years?

File: nff/nn/layers.py [hunk 4]
```diff
@@ -357,7 +357,8 @@ def __init__(self, l_spher, n_spher, cutoff, envelope_p):
         for i in range(l_spher):
             if i == 0:
                 first_sph = sym.lambdify([theta], self.sph_harm_formulas[i][0], modules)(0)
-                self.sph_funcs.append(lambda tensor: torch.zeros_like(tensor) + first_sph)
+                self.sph_funcs.append(partial(lambda tensor, fsph: torch.zeros_like(tensor) + fsph,
+                                              fsph=first_sph))
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/nn/layers.py:361]: True magic^^

File: nff/nn/modules/schnet.py [hunk 5]
```diff
@@ -1288,11 +1283,7 @@ def testBaseEdgeUpdate(self):
         r_in = torch.rand(6, 10)
         model = MessagePassingModule()
         r_out = model(r_in, e, a)
-        self.assertEqual(
-            r_in.shape,
-            r_out.shape,
-            "The node feature dimensions should be same for the base case",
-        )
+        assert r_in.shape == r_out.shape, "The node feature dimensions should be same for the base case"
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/nn/modules/schnet.py:1286]: `These methods are used instead of the assert statement so the test runner can accumulate all test results and produce a report.`

I guess it makes debugging while developing easier, but ruff does not like it, because it expects to be perfect^^


Moving the code to the new unittest folder like it did with some ASE based tests, would be good for the next clean up step (if the unittests still go through...)

[the remark this answers is not in the document]
Comment (ajhoffman1229) [reply nff/nn/modules/schnet.py:1286]: Ok, I will defer to you in this decision.

File: pyproject.toml [hunk 6]
```diff
@@ -124,5 +138,18 @@ ignore = [
     "S310",    # url open functions can be unsafe
     "TRY003",  # long exception messages not defined in the exception class itself
     "UP015",   # unnecessary "r" in open call
+    "Q000",    # single quotes
+    "D103",    # missing docstring in public function
+    "D",       # do not fight with doc for now, todo in the future
+    "UP035",   # deprecated Dict and List, but keep to support python<3.10
+    "UP006",   # deprecated Dict and List, but keep to support python<3.10
+    "ICN001",  # import name convention
+    "E741",    # 'l' as variable name
```
[the remark this answers is not in the document]
Comment (steinmig) [reply pyproject.toml]: Just a lot of occurrences :sweat_smile: 

 I think I know found a good compromise

File: pyproject.toml [hunk 7]
```diff
@@ -124,5 +138,18 @@ ignore = [
     "S310",    # url open functions can be unsafe
     "TRY003",  # long exception messages not defined in the exception class itself
     "UP015",   # unnecessary "r" in open call
+    "Q000",    # single quotes
+    "D103",    # missing docstring in public function
+    "D",       # do not fight with doc for now, todo in the future
+    "UP035",   # deprecated Dict and List, but keep to support python<3.10
+    "UP006",   # deprecated Dict and List, but keep to support python<3.10
```
[the remark this answers is not in the document]
Comment (steinmig) [reply pyproject.toml:145]: Not sure if this options breaks python<3.9

I can look into adding the option, when we add more python versions to the CI

File: pyproject.toml [hunk 8]
```diff
@@ -124,5 +138,18 @@ ignore = [
     "S310",    # url open functions can be unsafe
     "TRY003",  # long exception messages not defined in the exception class itself
     "UP015",   # unnecessary "r" in open call
+    "Q000",    # single quotes
```
[the remark this answers is not in the document]
Comment (steinmig) [reply pyproject.toml:141]: I don't remember if there was a particular reason for this. At least now it just sounds a bit risky to me for complex f-strings and is very nit-picky with no obvious advantage, so I would still allow single quotes

File: pyproject.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajhoffman1229) [reply pyproject.toml]: Oh jeez. People using `l` as a variable name is just crazy to me... Maybe once the broader codebase is ruff/flake8/mypy compliant, we can worry about this issue.

File: pyproject.toml [hunk 9]
```diff
@@ -124,5 +138,18 @@ ignore = [
...
+    "D103",    # missing docstring in public function
+    "D",       # do not fight with doc for now, todo in the future
```
[the remark this answers is not in the document]
Comment (steinmig) [reply pyproject.toml:143]: Agree, but the reason for excluding all D errors was also that ruff is very particulate what may or may not be in a docstring and how a docstring should look like, so I avoided reading through all the ruff docstring codes (and personal opinion: the rigidity rather encourages less documentation instead of better documentation)

File: nff/utils/xyz2mol.py [hunk 10]
```diff
@@ -12,60 +12,135 @@
...
+    "b",
+    "c",
+    "n",
+    "o",
+    "f",
+    "ne",
+    "na",
+    "mg",
+    "al",
+    "si",
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/utils/xyz2mol.py:49]: By now xyz2mol is also included in Rdkit, so maybe just get rid of it entirely :sweat_smile:

File: pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajhoffman1229) [reply pyproject.toml:145]: Ok, in principle the `keep-runtime-typing` shouldn't be affected by Python version because it will not flag code that chooses to use `typing` classes (`List`, `Tuple`, etc.) rather than the newer Python type annotation and checking objects (`list`, etc.). It should actually help for earlier versions of Python. I think it will also discourage the use of the `from __future__ import annotations` import, which is not compliant with earlier versions of Python.

Another option would be to eliminate support of Python versions < 3.9 or 3.10... I don't think there are any packages that NFF relies on that require these earlier versions of Python and can be run with Python 3.11 and higher. It might be worth encouraging that shift during this refactor since Python 3.9 will no longer be supported after this year.

File: nff/nn/models/schnet.py [hunk 11]
```diff
@@ -152,37 +137,29 @@ def convolve(self,
...
-    def forward(self,
-                batch,
-                xyz=None,
-                requires_stress=False,
-                **kwargs):
+    def forward(self, batch, xyz=None, requires_stress=False, **kwargs):
         """Summary
         Args:
-            batch (dict): dictionary of props
+            batch (dict): dictionary of propSchNetEdgeFilters
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/nn/models/schnet.py:162]:
```suggestion
            batch (dict): dictionary of props
```

true

File: nff/reactive_tools/nms.py [hunk 12]
```diff
@@ -1,97 +1,104 @@
...
+                * proj_vib_freq_cm_1[i]
+                * SPEEDOFLIGHT
+                * 100
+                * 2
+                * np.pi
+                * amplitude
+                * np.sin(2 * np.pi * random_0_1[i])
+                * proj_hessian_eigvec[i]
+                / Bohr**2
+            )
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/reactive_tools/nms.py:91]:
```suggestion
            velocities.append(
                -1
                * proj_vib_freq_cm_1[i]
                * SPEEDOFLIGHT
                * 100
                * 2
                * np.pi
                * amplitude
                * np.sin(2 * np.pi * random_0_1[i])
                * proj_hessian_eigvec[i]
                / Bohr**2
            )  # todo: properly import / name / document units and unit conversions
```

File: nff/reactive_tools/reactive_langevin.py [hunk 13]
```diff
@@ -1,75 +1,70 @@
+import numpy as np
 from ase.io import Trajectory
-from ase.md.langevin import *
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/reactive_tools/reactive_langevin.py:2]: Ah yes I didn't spot the unchecked call with a value of an input dictionary with a hard-coded key^^

I'll add it back until we cover the tutorial.

File: nff/utils/confs.py [hunk 14]
```diff
@@ -249,7 +238,7 @@ def update_nbr_idx_keys(dset, batch, i, old_nbrs, num_confs):
 
 
 def update_per_conf(dataset, i, old_num_atoms, new_n_confs):
-    mol_size = dataset.props["mol_size"][i]
+    dataset.props["mol_size"][i]
```
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/utils/confs.py:241]: see above

File: nff/utils/geom.py (hunk 1 above)
Comment (steinmig) [reply nff/utils/geom.py:236]: see above

File: pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (steinmig) [reply pyproject.toml:145]: makes sense

File: nff/md/colvars.py
[the remark this answers is not in the document]
Comment (steinmig) [reply nff/md/colvars.py:1]: I merge the changes from master

**Review by ajhoffman1229 — APPROVED** [review]
🚀

**Comment by simonaxelrod** [comment]
> Actually, after looking at this more closely, I am inclined to say that it should be deleted entirely. I think this is a test of one of Simon's models, but it's not obvious from the code what it's doing (I presume he had some configuration interaction calculations that he used to train a NFF?), it refers to a repo on which NFF should not depend at all, and has lots of broken code (missing/commented out imports, references to Simon's specific fork of htvs, references to Simon's personal folders on a workstation, etc.). It hasn't been updated in nearly 4 years and, from what I can tell, @xiaochendu was in the process of [deprecating it](https://github.com/learningmatter-mit/NeuralForceField/commit/aa24cd513178f24391bc06a6a00642d864517a5a) anyway. If @xiaochendu knows more, it would be great if he could chime in here!

This is one of the funniest takedowns I've ever seen hahahaha
